File: 979/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    move = 0

    def dfs(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        left = self.dfs(root.left)
        right = self.dfs(root.right)
        self.move += abs(left) + abs(right)
        return left + right + root.val - 1

    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return self.move

File: 979/solution.go
package _979

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var res int

func distributeCoins(root *TreeNode) int {
	res = 0
	dfs(root)
	return res
}

func abs(val int) int {
	if val < 0 {
		return -val
	}
	return val
}

func dfs(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := dfs(root.Left)
	right := dfs(root.Right)
	res += abs(left) + abs(right)
	return left + right + root.Val - 1
}
